### PR TITLE
[libc++] Use __wrap_iter in string_view and array in the unstable ABI

### PR DESCRIPTION
**Pull request by ldionne**
std::string_view and std::array iterators don't have to be raw pointers,
and in fact other implementations don't represent them as raw pointers.
Them being raw pointers in libc++ makes it easier for users to write
non-portable code. This is bad in itself, but this is even worse when
considering efforts like hardening where we want an easy ability to
swap for a different iterator type. If users depend on iterators being
raw pointers, this becomes a build break.

Hence, this patch enables the use of __wrap_iter in the unstable ABI,
creating a long term path towards making this the default. This patch
may break code that assumes these iterators are raw pointers for
people compiling with the unstable ABI.

This patch also removes several assumptions that array iterators are
raw pointers in the code base and in the test suite.

### CODE REVIEW

**Comment by ldionne** [comment]
CC @llvm/libcxx-vendors since this can break some code. CC @EricWF for awareness in particular.

**Review by philnik777 — APPROVED** [review]
LGTM.

File: libcxx/include/array
Comment (ldionne) [reply libcxx/include/array:175]: I'll do it separately for all these aliases.

File: libcxx/include/__config
Comment (cjdb) [inline libcxx/include/__config:172]: I'd suggest the motivation focus on the hardening opportunities. The portability aspect is nice, but framing this as potentially closing up security vulnerabilities makes this a lot more appealing.

**Review by cjdb — COMMENTED** [review]
I very much like this change for its hardening improvements. I'm going to check how much code this impacts internally right now: would it be possible to hold off merging this for a day or so, so that I can get details on what we need to do, please?

**Comment by cjdb** [comment]
> > I very much like this change for its hardening improvements. I'm going to check how much code this impacts internally right now: would it be possible to hold off merging this for a day or so, so that I can get details on what we need to do, please?
> 
> Yup, of course, the intent was to upload this and let folks see how much impact it has. Please let me know once you have some information. I think Google is going to be the main way we can get information here since most vendors ship the stable ABI.

Excellent, thank you! I'll give you an answer as soon as possible (likely tomorrow).

**Review by philnik777 — CHANGES_REQUESTED** [review]
That got a lot bigger. Maybe we should split the `array` and `string_view` changes?

**Comment by ldionne** [comment]
@cjdb I think this should be good to test now! Make sure to apply it on top of `main`, since it has one dependent patch.

**Comment by ldionne** [comment]
Pinging reviewers.

**Review by philnik777 — APPROVED** [review]

**Comment by ldionne** [comment]
@cjdb Do you think it'd be possible to take this for a run internally to see the impact? This is not a must for LLVM 18 but I'd like to unblock this.

**Comment by cjdb** [comment]
Sorry, I think this got lost in the haze that is December. I'll get an answer ASAP.

**Comment by ldionne** [comment]
Oh yeah, no worries! This is not an emergency, I just wanted to avoid forgetting this forever :)

**Comment by ldionne** [comment]
Gentle ping @cjdb , did you have any luck running this internally?

**Comment by ldionne** [comment]
I spoke to @cjdb offline who told me there was a non-insignificant number of build failures when enabling this. However, this change can easily be disabled temporarily (while the code base migrates) by not defining `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY` and `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW`.

Our ABI flags have that benefit that one can pick and choose what's opted into, so let's use this here. Hence, I think it is reasonable to move forward with the change, and vendors who envision breakage should simply disable these ABI macros until they are ready to transition.

CC @llvm/libcxx-vendors again

**Comment by rnk** [comment]
> However, this change can easily be disabled temporarily (while the code base migrates) by not defining `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY` and `_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW`.

Thanks for the migration pathway!

**Comment by steelannelida** [comment]
@rnk @ldionne could you please provide a more detailed way to work around this?

Is the suggestion to modify __config and undefine _LIBCPP_ABI_USE_WRAP_ITER_IN_STD_ARRAY and _LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW? Unfortunately this cannot be done in __config_site because the macros are defined after its inclusion.

**Comment by ldionne** [comment]
You can either modify `__config` to undefine those macros in the source, or you can start opting into individual ABI flags instead of opting into all of them by switching from `LIBCXX_ABI_VERSION = 2` (in CMake) to `LIBCXX_ABI_DEFINES` (in CMake). Depending on your use case, it may be simpler to just undefine the two macros locally for your transition period or to switch to a purely opt-in approach for ABI flags, but both work.

**Comment by cjdb** [comment]
If it's causing lots of pain, would it be possible to make this opt-in for two releases, then switch it to opt-out (and document that we'll be doing this no later than two releases time)? That'll give folks time to address this issue without being broken today.

To avoid having users need to either modify `__config` or need to maintain an unbounded set of macros, what do you think about flipping the macro? Then users who don't want/can't yet use this only need to add one macro instead of many.
```cpp
#if defined(_LIBCPP_ABI_BOUNDED_ITERATORS)
  using const_iterator = __bounded_iter<const_pointer>;
#elif defined(_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW)
  using const_iterator = __wrap_iter<const_pointer>;
#else
  using const_iterator = const_pointer;
#endif
```
to
```cpp
#if defined(_LIBCPP_ABI_BOUNDED_ITERATORS)
  using const_iterator = __bounded_iter<const_pointer>;
#elif !defined(_LIBCPP_ABI_NO_USE_WRAP_ITER_IN_STD_STRING_VIEW) // macro adds "NO"
  using const_iterator = const_pointer;
#else
  using const_iterator = __wrap_iter<const_pointer>;
#endif
```

Apologies for not thinking of this when we chatted last week.

**Comment by cjdb** [comment]
I have a patch that's slightly different to the forward-fix above, and I think lets us all win here.

**Comment by PiJoules** [comment]
> If it's causing lots of pain, would it be possible to make this opt-in for two releases, then switch it to opt-out (and document that we'll be doing this no later than two releases time)? That'll give folks time to address this issue without being broken today.
> 
> To avoid having users need to either modify `__config` or need to maintain an unbounded set of macros, what do you think about flipping the macro? Then users who don't want/can't yet use this only need to add one macro instead of many.
> 
> ```c++
> #if defined(_LIBCPP_ABI_BOUNDED_ITERATORS)
>   using const_iterator = __bounded_iter<const_pointer>;
> #elif defined(_LIBCPP_ABI_USE_WRAP_ITER_IN_STD_STRING_VIEW)
>   using const_iterator = __wrap_iter<const_pointer>;
> #else
>   using const_iterator = const_pointer;
> #endif
> ```
> 
> to
> 
> ```c++
> #if defined(_LIBCPP_ABI_BOUNDED_ITERATORS)
>   using const_iterator = __bounded_iter<const_pointer>;
> #elif !defined(_LIBCPP_ABI_NO_USE_WRAP_ITER_IN_STD_STRING_VIEW) // macro adds "NO"
>   using const_iterator = const_pointer;
> #else
>   using const_iterator = __wrap_iter<const_pointer>;
> #endif
> ```
> 
> Apologies for not thinking of this when we chatted last week.

Do you have a PR for something like this? I think it would also be nice to have a macro we can set in our build that doesn't involve editing the `__config` file or opting out of `LIBCXX_ABI_VERSION = 2`.

**Comment by ldionne** [comment]
@PiJoules Please see https://github.com/llvm/llvm-project/pull/84226#issuecomment-1981834012.

**Pull request merged**
